Feature gate clean

LeoTestard opened the pull request:
This PR does a bit of cleaning in the feature-gate-handling code of libsyntax. It also fixes two bugs (#32782 and #32648). Changes include:

* Change the way the existing features are declared in `feature_gate.rs`. The array of features and the `Features` struct are now defined together by a single macro. `featureck.py` has been updated accordingly. Note: there are now three different arrays for active, removed and accepted features instead of a single one with a `Status` item to tell wether a feature is active, removed, or accepted. This is mainly due to the way I implemented my macro in the first time and I can switch back to a single array if needed. But an advantage of the way it is now is that when an active feature is used, the parser only searches through the list of active features. It goes through the other arrays only if the feature is not found. I like to think that error checking (in this case, checking that an used feature is active) does not slow down compilation of valid code. :) But this is not very important...
* Feature-gate checking pass now use the `Features` structure instead of looking through a string vector. This should speed them up a bit. The construction of the `Features` struct should be faster too since it is build directly when parsing features instead of calling `has_feature` dozens of times.
* The MacroVisitor pass has been removed, it was mostly useless since the `#[cfg]-stripping` phase happens before (fixes #32648). The features that must actually be checked before expansion are now checked at the time they are used. This also allows us to check attributes that are generated by macro expansion and not visible to MacroVisitor, but are also removed by macro expansion and thus not visible to PostExpansionVisitor either. This fixes #32782. Note that in order for `#[derive_*]` to be feature-gated but still accepted when generated by `#[derive(Trait)]`, I had to do a little bit of trickery with spans that I'm not totally confident into. Please review that part carefully. (It's in `libsyntax_ext/deriving/mod.rs`.)::

Note: this is a [breaking change], since programs with feature-gated attributes on macro-generated macro invocations were not rejected before. For example:

```rust
macro_rules! bar (
    () => ()
);

macro_rules! foo (
    () => (
        #[allow_internal_unstable] //~ ERROR allow_internal_unstable side-steps
        bar!();
    );
);
```
foo!();

rust-highfive commented:
r? @nrc

(rust_highfive has picked a reviewer for you, use r? to override)

eddyb commented:
cc #32655 and #32671

LeoTestard commented:
Ha, I see we used the same trick for spans for `#[derive_*]`. :) This reassures me a bit.

eddyb commented:
@LeoTestard AFAICT syntex was fixed not to generate code that trips over such a check, but @nikomatsakis [would still like a warning for a release cycle](https://github.com/rust-lang/rust/pull/32671#issuecomment-206245303) (old syntex output might still be around).

LeoTestard commented:
You mean (well, he means) generate a warning instead of an error? That sounds tricky with my approach since the check in `libsyntax_ext/deriving/mod.rs` now not only catches uses inside macros but also all other uses. I have no idea how to check if a use comes from a macro or not. Suggesting to run `cargo update` when the error comes from syntex sounds feasible though.

eddyb commented:
@LeoTestard the error is not "from syntex", but rather you get a file that syntex pretty-printed. I can't think of a way to distinguish that, since the actual code could be anything the user put through syntex.

LeoTestard commented:
Ok now it should only *warn* for `#[derive_*]` for the cases that did not error before. It's tricky and based on whether the parent of the top of the expansion backtrace points to a valid `ExpnInfo` and is not `NO_EXPANSION`. It seems to work for the few interesting cases I've tested.

Note that this PR is still a breaking change for all the other uses of feature-gated attributes that were not checked in macro expansion (see #32782) and that are now checked and just reported as errors.

r?

bors commented:
:umbrella: The latest upstream changes (presumably #32016) made this pull request unmergeable. Please resolve the merge conflicts.

LeoTestard commented:
Rebased onto master. r?
(cc @nikomatsakis @pnkfelix)

jseyfried commented:
@LeoTestard I believe this PR is only feature gate checking attributes of macros in item positions.

For example, the following would be erroneously allowed (it is an error now):
```rust
macro_rules! method {
    () => { fn f() {} }
}

struct S;
impl S {
    #[allow_internal_unstable]
    method! {}
}
```

jseyfried commented on src/libsyntax_ext/deriving/mod.rs:
```diff
@@ -96,6 +96,36 @@ fn expand_derive(cx: &mut ExtCtxt,
 
+            // This span is **very** sensitive and crucial to
+            // getting the stability behavior we want. What we are
+            // doing is marking `#[structural_match]` with the
```
If I understand correctly, we could also be marking `#[derive_*]` with this span -- the comment should probably reflect that.

jseyfried commented on src/librustc_driver/driver.rs:
```diff
@@ -515,9 +515,7 @@ pub fn phase_2_configure_and_expand(sess: &Session,
     time(time_passes, "gated macro checking", || {
```
This isn't doing gated macro checking anymore. Also, I don't think `syntax::feature_gate::get_features` needs its own timed pass since it is doing so little.
Perhaps we could time a single task consisting of `get_features` and the preceding three tasks (`update_recursion_limit`, `compute_crate_disambiguator`, and `collect_crate_types`, the latter two of which are not currently timed)?

jseyfried commented on src/libsyntax/feature_gate.rs:
```diff
@@ -1028,22 +987,22 @@ impl<'a, 'v> Visitor<'v> for PostExpansionVisitor<'a> {
+                gate_feature_post!(&self, box_syntax,
                                   e.span,
                                   "box expression syntax is experimental; \
                                    you can call `Box::new` instead.");
```
nit: this can be just `gate_feature_post!(&self, box_syntax, e.span, EXPLAIN_BOX_SYNTAX);`

jseyfried commented on src/libsyntax/feature_gate.rs:
```diff
@@ -661,57 +665,58 @@ const EXPLAIN_STMT_ATTR_SYNTAX: &'static str =
         return;
     }
     emit_feature_err(diag, "box_syntax", span, GateIssue::Language, EXPLAIN_BOX_SYNTAX);
 }
```
`check_for_box_syntax` can be removed entirely (it is unused)

eddyb replied:
I think that there's a number of `check_*` functions in that module which are dead code.

jseyfried commented:
Reviewed, LGTM modulo comments.

LeoTestard commented:
@jseyfried Ok, I fixed what you said. I added another commit for removing useless code since it's not directly related. Let me know if you want me to squash it/rebase it/whatever.

jseyfried commented on src/libsyntax/ext/expand.rs:
```diff
@@ -70,10 +70,13 @@ pub fn expand_expr(e: P<ast::Expr>, fld: &mut MacroExpander) -> P<ast::Expr> {
+            if let Some(&Features { pushpop_unsafe: false, .. }) =
+                fld.cx.ecfg.features {
+                feature_gate::emit_feature_err(
+                    &fld.cx.parse_sess.span_diagnostic, "pushpop_unsafe", expr_span,
```
This should be `"placement_in_syntax"`.

LeoTestard replied:
Yeah I saw that but I didn't have the time to fix it earlier... I copy-pasted too fast. Sorry about that.

jseyfried replied:
No prob!

jseyfried commented on src/libsyntax/ext/expand.rs:
```diff
@@ -70,10 +70,13 @@ pub fn expand_expr(e: P<ast::Expr>, fld: &mut MacroExpander) -> P<ast::Expr> {
-                &fld.cx.parse_sess.span_diagnostic,
-                expr_span);
+            if let Some(&Features { pushpop_unsafe: false, .. }) =
+                fld.cx.ecfg.features {
```
Could this be `if fld.cx.ecfg.features.unwrap().placement_in_syntax {`?
I think it would be better to panic if `features` were erroneously `None` for some reason than to silently bypass gated feature checks.

jseyfried commented on src/libsyntax/ext/expand.rs:
```diff
@@ -367,6 +367,14 @@ pub fn expand_item_mac(it: P<ast::Item>,
+                                      &fld.cx.parse_sess.codemap(),
+                                      &fld.cx.ecfg.features.unwrap());
+    }
+
```
I believe we still need to stability check macros in expression and statement positions (search for `drop(attrs)` in this file).
I would refactor this out into a new function in `expand.rs` (something like `check_attributes(attr: &ast::Attribute, fld: &mut MacroExpander)`) and then inline and remove `feature_gate::check_attribute`.

LeoTestard replied:
Are you sure we should inline? `feature_gate::Context` is private for the moment...

jseyfried replied:
Good point -- not inlining is ok

jseyfried commented on src/libsyntax/ext/expand.rs:
```diff
@@ -367,6 +376,10 @@ pub fn expand_item_mac(it: P<ast::Item>,
     });
 
+    // check here to detect use of feature-gated or invalid attributes on
+    // macro invoations since they will not be detected after macro expansion
```
nit: move this comment to `fn check_attributes` ("this function is called to detect use ...")

jseyfried commented:
@LeoTestard could you add that this is a [breaking-change] in the initial comment and give an example that breaks?

LeoTestard commented:
Done.

jseyfried commented:
@LeoTestard thanks! This looks good to me (modulo typos causing travis to fail).

I think this is likely enough to cause breakage in practice to warrant a crater run (cc @alexcrichton).
r=me if the crater run is clean.

LeoTestard commented:
Oooops. Looks like I fucked up the git history of this branch a little. Let me fix that...

LeoTestard commented:
Okay, looks fine now.  Let's wait for the crater run then. :)

bors commented:
:umbrella: The latest upstream changes (presumably 49be3dd) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
FYI I started a crater run just now.

nrc commented on src/librustc_driver/driver.rs:
```diff
@@ -512,19 +512,13 @@ pub fn phase_2_configure_and_expand(sess: &Session,
+    sess.track_errors(|| {
+        *sess.features.borrow_mut() =
+            syntax::feature_gate::get_features(&sess.parse_sess.span_diagnostic,
+                                            &krate);
```
nit: indentation

nrc commented:
Does this remove feature gate checking for macros? It seems that feature gates for features which are expanded away during macro expansion won't be checked at all (e.g., `trace_macros!`). That seems bad.

nrc commented on src/libsyntax/feature_gate.rs:
```diff
@@ -259,16 +271,47 @@ const KNOWN_FEATURES: &'static [(&'static str, &'static str, Option<u32>, Status
+    ($field: ident) => {{
+        fn f(features: &Features) -> bool {
+            features.$field
+        }
```
Would it be possible to make `Features` an array rather than a struct and index into it using a constant for each feature, rather than this combination of function and macro? Or do we sometimes use a function other than the macro-generated one?

jseyfried replied:
> Or do we sometimes use a function other than the macro-generated one?

I believe we do not. I kind of like that this implementation does not have hard-coded indices, but it would be worth it to avoid the redundant `cfg_fn!`s (although it would require mildly refactoring all the feature checks scatted throughout the compiler).

LeoTestard replied:
I thought of using an array but I found the approach with a structure to be safer (it avoids hard-coded indices though generating them with a macro would not be that bad). The `cfg_fn!`s are there mostly to find a feature by name from a string, which is useful for two cases: building the structure from the crate attributes, and checking an attribute (you have the attribute as a string and you must get the corresponding structure field). More simply: it allows you to check (or set) a feature whose name is not known at compile-time.

nrc commented on src/libsyntax/feature_gate.rs:
```diff
@@ -682,36 +817,37 @@ pub fn check_for_pushpop_syntax(f: Option<&Features>, diag: &Handler, span: Span
-            emit_feature_err(self.span_handler, feature, span, GateIssue::Language, explain);
+macro_rules! gate_feature_fn {
+    ($cx: expr, $has_feature: expr, $span: expr, $name: expr, $explain: expr) => {{
+        let (cx, has_feature, span, name, explain) = ($cx, $has_feature, $span, $name, $explain);
```
What is this line for?

This whole macro makes me feel bad, it might be necessary, but I'd love to see it done more simply.

LeoTestard replied:
Previously we had a `gate_feature` feature function that took a feature name as a string, an span, and an error message and would check that the feature was enabled, and, if not, report an error message at that span. Now that feature names are attributes, it must be a macro. This specific line is to avoid duplication of expressions. It might not be necessary in this specific case since all the invocations of this macro are pretty simple, but I have the habit of doing that when I write macros. In more complex cases, not doing so might result in duplicating computation (which can be very bad in the presence of side effects) or in the case of Rust, borrow-checker errors, I guess.

jseyfried replied:
@LeoTestard couldn't you make `gate_feature_fn` an ordinary function instead of macro? e.g.
```rust
fn feature_gate_fn<F>(cx: &Context, has_feature: F, span: Span, name: &str, explain: &str)
    where F: Fn(&Features) -> bool
{ ... }
```

jseyfried commented:
> Does this remove feature gate checking for macros?

No. For example, `trace_macros!` is still gated feature checked [here](https://github.com/LeoTestard/rust/blob/f9ba1078245bc2c023f51d2a63e0ca84863600e7/src/libsyntax_ext/trace_macros.rs#L23-L30).

nrc commented:
> No. For example, trace_macros! is still gated feature checked here.

That is good. Why is the test for it removed? Or is it moved and I missed the destination?

jseyfried commented:
I think @LeoTestard considered it to be redundant with [this test](https://github.com/LeoTestard/rust/blob/feature-gate-clean/src/test/compile-fail/trace_macros-gate.rs) -- both test that we gated feature check macro expanded `trace_macros!` invocations.

LeoTestard commented:
> I think @LeoTestard considered it to be redundant with this test -- both test that we gated feature check macro expanded trace_macros! invocations.

Indeed. There were previously three tests for checking of feature-gated macros because some uses were detected before expansion and some during expansion. The compiler would abort if there were errors during the first (pre-expansion) check, so we needed several tests: one that checked that the first pass errored, one that checked that if the first succeeded, the second errored, and one that checked that if the first errored, the second doesn't. Now all errors are reported at once so a single test case is enough.

nikomatsakis commented:
It seems that the PR rebased since I started the crater run, but I think these were the results:

https://gist.github.com/nikomatsakis/9f974b9da1fdbe01263335e61426869c

(I had a few runs going, so I might be wrong, since I usually correlate later via the SHA1 hashes.)

nikomatsakis commented:
Summary: 4584 crates tested: 2876 working / 1354 broken / 95 regressed / 29 fixed / 230 unknown.

With 14 root regressions.

eddyb commented:
@LeoTestard looks like you need to let the expansion work when outputting the warning, otherwise you get errors from missing implementations.

LeoTestard commented:
@eddyb Ha yes, indeed, fixed that. Some other regressions don't look related and by ckecking the commits, I see that there are a lot more commits between mine and the reference, probably because of the rebases. Let's do another crater run...

bors commented:
:umbrella: The latest upstream changes (presumably #32803) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
What's the status on this? I was surprised to find #32671 still open, I thought I saw this getting r+'d.

LeoTestard commented:
@eddyb We were supposed to do another crater run with the code amended to continue expansion even when the warning is emitted (and without including unrelated commits that were responsible of some breakage in the first run) and see how it goes but as far as I know @nikomatsakis had trouble running it and I don't know what the current status is.

I have a rebased version on my work computer (although it will probably need to be rebased again because it was last updated a week ago or so), I'll push it tomorrow. But I guess we can still run the crater test with this version in the meantime.

nikomatsakis commented:
@LeoTestard crate run <https://gist.github.com/b903ceb777c47156ce2a89c3b465d29c> shows:

* 4349 crates tested: 2807 working / 1215 broken / **1 regressed** / 0 fixed / 326 unknown.

* [uri-0.1.1](https://crates.io/crates/uri) ([before](https://tools.taskcluster.net/task-inspector/#ke66b8LuTZeuPKKlXWQUFg)) ([after](https://tools.taskcluster.net/task-inspector/#SnNB4HYPTDCm3sLrTR0yqg))

nikomatsakis commented:
Moreover, that looks like a false positive. You could consider checking it locally, but it's probably fine.

LeoTestard commented:
@nmataskis Tested locally and it compiles without problem. Just need to rebase now...

LeoTestard commented:
Done. r?

nikomatsakis commented:
@LeoTestard it seems like tidy doesn't build on this branch?

LeoTestard commented:
@nikomatsakis Huh, yeah, my bad, I thought I tested this but that was before I had to rebase it against the complete rewriting of tidy checks I guess. u__u

LeoTestard commented:
@nikomatsakis Fixed, I think.

nikomatsakis commented:
r? @nikomatsakis

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit fa23d10 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis fa23d108638b7117e84c99a8d36a9dbab7b56812 -->

bors commented:
:hourglass: Testing commit fa23d10 with merge 435095f...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/8955), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/8990), [auto-linux-32cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-32cross-opt/builds/79), [auto-linux-64-cargotest](http://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/271), [auto-linux-64-cross-armhf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armhf/builds/79), [auto-linux-64-cross-armsf](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-armsf/builds/82), [auto-linux-64-cross-freebsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/83), [auto-linux-64-cross-netbsd](http://buildbot.rust-lang.org/builders/auto-linux-64-cross-netbsd/builds/82), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/2222), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/8929), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/8917), [auto-linux-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/495), [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/894), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/8912), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/2124), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/4069), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/8963), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8974), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/8934), [auto-mac-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/895), [auto-mac-cross-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/82), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3994), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/4009), [auto-win-gnu-32-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/899), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/4005), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3983), [auto-win-msvc-32-cross-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-cross-opt/builds/365), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3202), [auto-win-msvc-64-cargotest](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/277), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/3995), [auto-win-msvc-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/498), [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/907)
